vyper/venom/analysis/loop_detection.py

```diff
@@ -0,0 +1,70 @@
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.analysis import IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.basicblock import IRBasicBlock
+
+
+class NaturalLoopDetectionAnalysis(IRAnalysis):
+    """
+    Detects loops and computes basic blocks
+    and the block which is before the loop
+    """
+
+    # key = start of the loop (last bb not in the loop)
+    # value = all the block that loop contains
+    loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+
+    done: OrderedSet[IRBasicBlock]
+    visited: OrderedSet[IRBasicBlock]
+
+    def analyze(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]] = dict()
+        self.done = OrderedSet()
+        self.visited = OrderedSet()
+        entry = self.function.entry
+        self._dfs_r(entry)
+
+    def _dfs_r(self, bb: IRBasicBlock, before: IRBasicBlock | None = None):
+        if bb in self.visited:
+            self.done.add(bb)
+            if before is None:
+                return
+            loop = self._collect_path(before, bb)
+            in_bb = bb.cfg_in.difference({before})
```

Comment:
i think this might be clearer as `if before in bb.cfg_in: ...`. does that introduce any bugs?

Reply:
This is just because it checks for  natural loops, so I check if it has only one input into the loop.